### PR TITLE
Add NTP zgrab2 module

### PR DESCRIPTION
**Pull request by justinbastress**
NTP module for zgrab2. Supports both normal operation and sending MONLIST requests.

## How to Test

Normal operation can be tested against any NTP server, e.g. `echo "time.nist.gov" | cmd/zgrab2/zgrab2 ntp --timeout 3 | jp data`. The timestamp should appear in `data.ntp.result.time`.

To test the `--monlist` mode, you obviously will need an NTP service with MONLIST enabled -- the container in `integration-test/ntp/container-4.2.6` can serve this purpose (it builds [NTP 4.2.6](https://www.eecis.udel.edu/~ntp/ntp_spool/ntp4/ntp-4.2/ntp-4.2.6p5.tar.gz) with the default options, which allows MONLIST). So you could run

 * `docker build -t zgrab_ntp:4.2.6 integration_tests/ntp/container-4.2.6`
 * `docker run -itd -p 33123:123/udp --rm --name ntptest zgrab_ntp:4.2.6`
 * `echo "127.0.0.1" | cmd/zgrab2/zgrab2 ntp --port 33123 --timeout 3 --monlist`

There should be a `data.ntp.result.monlist_response` with some binary data and a `data.ntp.result.monlist_header` with more details on the response, in addition to the `data.ntp.result.time` and `data.ntp.result.time_response` fields that were present before (they should disappear if you pass `--skip-get-time`).

## Notes & Caveats

 * For some reason, on my machine at least, forwarding UDP to ::1 on docker containers does not seem to work (forwarding to IPv4 addresses does, and forwarding TCP to IPv6 addresses works as well).

## Issue Tracking

Part of the DDoS amplifiers list

### CODE REVIEW

**Comment by zakird** [comment]
It's great to see this added to ZGrab. One quick question, to find NTP
amplifiers, do we need to do a stateful scan (e.g. pipe ZMap into ZGrab) or
can this be done in a stateless ZMap scan alone?

**Comment by justinbastress** [comment]
@zakird as it stands, it's able to do the UDP connection and send the MONLIST packet from within in the existing zgrab2 framework -- the NTP module just interprets the port to mean "udp port", which works because each module sets up its own connections (as far as zgrab2 is concerned, we could be ignoring the port number / hostname completely).

File: modules/ntp.go [hunk 1]
```diff
@@ -0,0 +1,844 @@
...
+		return nil, err
+	}
+
+	return &ret, nil
+}
+
+// readNTPHeader() reads 48 bytes from conn and interprets it as an NTPHeader
+func readNTPHeader(conn net.Conn) (*ntpHeader, error) {
+	buf := make([]byte, 48)
+	_, err := io.ReadFull(conn, buf)
```
Comment (dadrian) [inline modules/ntp.go]: Yay, UDP let's us do this safely!

**Comment by dadrian** [comment]
I haven't gone digging into the details of this too much yet, but here's a couple high level comments:
- We should avoid returning private structs from public functions, and avoid making public fields on structs be private types. I'll see if I can dig up a good article on this.
- We should start putting a brief description of the protocol and how the scanner scans the protocol as a package-level docs for the module. Example below, only needs to be in one file.

```
// Package sort provides primitives for sorting slices and user-defined
// collections.
package sort
```

File: modules/ntp.go [hunk 2]
```diff
@@ -0,0 +1,844 @@
...
+
+// infoError is a 3-bit integer, values taken from ntp_request.h
+type infoError uint8
+
+const (
+	infoErrorOkay     infoError = 0
+	infoErrorImpl               = 1
+	infoErrorReq                = 2
+	infoErrorFmt                = 3
+	infoErrorNoData             = 4
```
[the remark this answers is not in the document]
Comment (justinbastress) [reply modules/ntp.go]: Fixed.

**Pull request merged**